linera-chain/src/data_types.rs

```diff
@@ -421,17 +422,56 @@ pub struct LiteValue {
 struct ValueHashAndRound(CryptoHash, Round);
 
 /// A vote on a statement from a validator.
-#[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct Vote {
-    pub value: HashedCertificateValue,
+#[derive(Clone, Debug, Serialize)]
+pub struct Vote<T> {
+    pub value: Hashed<T>,
     pub round: Round,
     pub validator: ValidatorName,
     pub signature: Signature,
 }
 
-impl Vote {
+impl<'de, T: DeserializeOwned + BcsHashable> Deserialize<'de> for Vote<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Vote<T>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Debug, Deserialize)]
+        struct VoteHelper<T> {
```

Comment:
This is going to make `VoteHelper` appear in the recorded formats, disregarding `T`. You need to use `serde_name` for this.

Reply:
It's also a potential security vulnerability because we use the `serde`-provided name as a seed for the hashing, but here the seed won't take `T` into account here. It looks like the format is not recorded (because we're converting votes into legacy RPC messages?) so maybe it's not a blocker for this PR but you need to add `TODO(#nnn)` with a task and we need to start doing the right thing asap.

Reply:
This is `Deserialize` instance, not `Serialize` (that one is still derived) so it doesn't appear in the logs? 

> because we're converting votes into legacy RPC messages?

We shouldn't 🤔 I haven't added any conversions for that IIRC.

Reply:
Point taken for `Deserialize`. Curious what this is used for, though

Reply:
Deserializing a `Hashed<T>` adds a new trait bound `T: BcsHashable` – b/c `Hashed<_>` type does not serialize the hash, just the value, and we recreate the hash when deserializing. I didn't add this bound on the `Vote<T>` struct itself for two reasons:
1) trait bounds on structs are not good
2) compiler complains about it anyway (see [here](https://github.com/linera-io/linera-protocol/pull/2887/files#r1841937994) for details)

Reply:
I realized that I actually can do this with derive but I need additional attributes - https://github.com/linera-io/linera-protocol/pull/2887/commits/b4e8e64d208ddef2025be12f8031485aa2dcf4d3